File: bindings/java-jni/src/main/java/org/maplibre/nativejni/log/LogSeverity.java
package org.maplibre.nativejni.log;

/** Severity for a Maplibre Native log record. */
public final class LogSeverity {
  public static final LogSeverity INFO = new LogSeverity(1, "INFO");
  public static final LogSeverity WARNING = new LogSeverity(2, "WARNING");
  public static final LogSeverity ERROR = new LogSeverity(3, "ERROR");

  private final int nativeValue;
  private final String name;

  private LogSeverity(int nativeValue, String name) {
    this.nativeValue = nativeValue;
    this.name = name;
  }

  public int nativeValue() {
    if (name == null) {
      throw new IllegalArgumentException("Unknown log severity cannot be used as an input");
    }
    return nativeValue;
  }

  public int nativeMask() {
    return 1 << nativeValue();
  }

  public int rawValue() {
    return nativeValue;
  }

  public static LogSeverity fromNative(int nativeValue) {
    return switch (nativeValue) {
      case 1 -> INFO;
      case 2 -> WARNING;
      case 3 -> ERROR;
      default -> new LogSeverity(nativeValue, null);
    };
  }

  @Override
  public boolean equals(Object other) {
    return other instanceof LogSeverity value && nativeValue == value.nativeValue;
  }

  @Override
  public int hashCode() {
    return Integer.hashCode(nativeValue);
  }

  @Override
  public String toString() {
    return name != null ? name : "LogSeverity(" + nativeValue + ")";
  }
}

File: bindings/kotlin-native/src/nativeMain/kotlin/org/maplibre/nativeffi/log/LogSeverity.kt
package org.maplibre.nativeffi.log

import kotlin.jvm.JvmInline

/** Severity for a Maplibre Native log record. */
@JvmInline
public value class LogSeverity(public val nativeValue: Int) {
  public companion object {
    public val INFO: LogSeverity = LogSeverity(1)
    public val WARNING: LogSeverity = LogSeverity(2)
    public val ERROR: LogSeverity = LogSeverity(3)

    internal fun fromNative(nativeValue: UInt): LogSeverity = fromNative(nativeValue.toInt())

    internal fun fromNative(nativeValue: Int): LogSeverity = LogSeverity(nativeValue)
  }

  public val nativeMask: Int
    get() {
      require(isKnown) { "Unknown log severity cannot be used as an input: $nativeValue" }
      return 1 shl nativeValue.toInt()
    }

  internal val isKnown: Boolean
    get() = this == INFO || this == WARNING || this == ERROR
}
